main/testkit/src/testkit/MillTestkit.scala

```diff
@@ -0,0 +1,152 @@
+package mill.testkit
+
+import mill.define._
+import mill.api.Result
+import mill.api.Result.OuterStack
+import mill.api.Strict.Agg
+import java.io.{InputStream, PrintStream}
+import mill.define.{Input, Target, Task}
+import mill.eval.Evaluator
+import language.experimental.macros
+import mill.api.{DummyInputStream, Result}
+
+import scala.collection.mutable
+
+class MillTestKit(targetDir: os.Path) {
```

Comment:
> I wonder whether it is a better idea to make targetDir a def inside the class, as we do with other settings like externalOutPath? This would make it more flexible when inherited in an object and would allow some kind of dynamic default.

Sure, I'm cool with that. What would you like the default impl to be ? Something like 

```scala
def targetDir: os.Path = 
  sys.env.get("MILL_TESTKIT_BASEDIR")
       .map(convertToPath) 
       .getOrElse(os.pwd / "out" / "something_with_a_low_collision_risk") ?  
```

Also, is there a built-in function to parse a string into a `RelPath` ?

Reply:
Yeah, something like that. Don't know what `convertToPath` is, but `Path.apply(_, os.pwd)` can be used to create a `os.Path` from a string. And `os.rel / aString` can be used to create a `os.RelPath`.

As a fallback, maybe we can create a temp dir with `os.temp.dir`. Either in the system temp dir or under `os.pwd / "target"`?

If we need to read or copy resources in these tests, I nowadays prefer to read them from the classpath. This makes the test independent of the executing build tool and operating system.

Reply:
> If we need to read or copy resources in these tests, I nowadays prefer to read them from the classpath. This makes the test independent of the executing build tool and operating system. 

I think the `target` is ore about where mill's `out` folder gets saved, so it's a "writing" concern more than it is a "reading" one. 

Falling back to `os.temp.dir` makes sense :+1:

Reply:
Just a short note, as I saw this code snipped in a second review.

> ```scala
> def targetDir: os.Path = 
>   sys.env.get("MILL_TESTKIT_BASEDIR")
>        .map(convertToPath) 
>        .getOrElse(os.pwd / "out" / "something_with_a_low_collision_risk") ?  
> ```

If I really need dedicated temporary/permanent storage to be shared by more that one target in Mill, I typically end up creating a dedicated persistent target for them. As an example, the target directories we use when we generate IntelliJ IDEA project files, use this.

```scala
// taken from GenIdeaModule.scala
  def ideaCompileOutput: T[PathRef] = T.persistent {
    PathRef(T.dest / "classes")
  }
```